numpy/core/src/umath/ufunc_object.c

```diff
@@ -3634,7 +3634,8 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
         }
     } else if(Py_TYPE(initial) != Py_TYPE(identity)) {
         PyErr_Format(PyExc_TypeError,
-                    "initial type does not match the array type");
+                    "initial type %s does not match the array type %s",
+                    initial->ob_type->tp_name,identity->ob_type->tp_name);
```

Comment:
Since I randomly saw this. `PyErr_Format` also supports `%S` , and `ob_type` access would typically be spelled `Py_TYPE(initial)`.

The complexity here is that `identity` is actually just a generic type. We will probably need to convert `initial` to a NumPy array. After that you can use `PyArray_CanCastTo(..., ..., NPY_SAME_KIND_CASTING)`. To make things quicker for NumPy scalars, we could optimize the common cases, but I think a comment to that effect will do for now. The functions to do that are just private and in a bit of a turmoil in my opinion.